include/ccf/endpoints/authentication/authentication_types.h

```diff
@@ -35,7 +35,7 @@ namespace ccf
       std::string& error_reason) = 0;
 
     virtual void set_unauthenticated_error(
-      std::shared_ptr<ccf::RpcContext>& ctx, std::string&& error_reason);
+      std::shared_ptr<ccf::RpcContext> ctx, std::string&& error_reason);
```

Comment:
With `&`, this needs to be passed an lvalue shared ptr to an actual `RpcContext`, which may be modified. We don't want to modify it  (not, we might modify `*ctx`, but we don't modify `ctx` itself). The caller has a shared ptr to a derived type. Passing by value or `const&` is fine, but not `&`.

Specifically, the caller has a `std::shared_ptr<http::HttpRpcContext>`. That can't bind to an lvalue `&`, because that allows the implementation to call something like `ctx = std:;make_shared<OtherRpcContext>`, which means the caller now has junk.